### ISSUE TITLE
Question: Working with submodules

### ISSUE THREAD

**Issue by dmathewwws** [aws/copilot-cli#2989]
Hi, thanks for the useful project. It's made working with AWS a lot less overwhelming for me. 😄

I do have a question around adding a pipeline when my project includes a submodule.

I followed the instructions in #2799 about this issue, but I think im missing something.

1) Updated my `pipeline.yml` file to add output_artifact_format: CODEBUILD_CLONE_REF
```
  properties:
    branch: main
    repository: MY_REPO_URL
    output_artifact_format: CODEBUILD_CLONE_REF
```

2) Ran `copilot pipeline update`

3) Adjusted `buildspec.yml`  to add env git-credential-helper: yes AND added the submodule commands for my submodule (named shared)

```
version: 0.2
env:
  git-credential-helper: yes
phases:
  install:
    runtime-versions:
      docker: 18
      ruby: 2.6
    commands:
      - echo "cd into $CODEBUILD_SRC_DIR"
      - cd $CODEBUILD_SRC_DIR
      - git config --file=.gitmodules submodule.shared.url https://git-codecommit.us-west-2.amazonaws.com/v1/repos/shared
      - cat .gitmodules
      - git submodule update --init
```

I'm running into a 403 forbidden error that makes me think that `https://git-codecommit.us-west-2.amazonaws.com/v1/repos/shared` isn't the right url to add there, I've tried some different combination but not sure what I'm supposed to add to that line.

Thanks for you help!

**Comment by Lou1415926** [comment]
Hello @dmathewwws!

Your `buildspec.yml` looks good to me. This is likely because the `BuildProjectRole` lacks permissions to achieve what you want. 

The CodeBuild project assumes the `BuildProjectRole` and makes request on your behalf. If it doesn't have the permission required - which is `codecommit:GitPull` - in order to add a CodeCommit repository as a submodule, you will see the 403 error.

To add the additional permission, you can:
#### Go to the CodeBuild console
1. Look for the project named `pipeline-<app name>-<pipeline name>-BuildProject`. 
2. Go to the "Build details" tab. 
3. Under the "Environment" section, look for "Service role", there should be a link to the service role, which is exactly the `BuildProjectRole` that I've mentioned above. 
#### After clicking the link, you are brought to the IAM console
1. Select "Add inline policy"
2. Select "JSON" tab
3. Copy and paste the following snippet:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": "codecommit:GitPull",
            "Resource": <Replace this with the ARN for your CodeCommit repository>
        }
    ]
}
```
After this, you can trigger another build and see if it works!

**Comment by Lou1415926** [comment]
This issue so far could be possibly related to #1689 and #2755. We might need to develop a good way to natively support customizing `BuildProjectRole`.

If the above solution works for you, would you please give those two issues a thumb-up? This would help us prioritize the task!

**Comment by dmathewwws** [comment]
Thanks Wanxian!

I was able to follow your instructions, go to the IAM console and add the `codecommit:GitPull` permission. 

However I still do see the below error when I look at my build logs:
```Error retrieving Git credentials.no error: GITHUB Git credential unavailable.```

I was wonder if:
- it mattered that I am not using CodeCommit to host my repository, I am hosting the submodule on GitHub. Does that mean I need to set a different permission?

- I've added this line inside my `buildspec.yml`
```
git config --file=.gitmodules submodule.${SUBMODULE_NAME}.url https://git-codecommit.us-west-2.amazonaws.com/v1/repos/${GITHUB_USERNAME}/${SUBMODULE_NAME}
```

I guessed the url structure based on the comment in #2799. I am wondering if you think that line needed or accurate?

this is the whole first part of the `buildspec.yml` for context:
```
version: 0.2
env:
  git-credential-helper: yes
phases:
  install:
    runtime-versions:
      docker: 18
      ruby: 2.6
    commands:
      - echo "cd into $CODEBUILD_SRC_DIR"
      - cd $CODEBUILD_SRC_DIR
      - git config --file=.gitmodules submodule.${SUBMODULE_NAME}.url https://git-codecommit.us-west-2.amazonaws.com/v1/repos/${GITHUB_USERNAME}/${SUBMODULE_NAME}
      - cat .gitmodules
      - git submodule update --init
```

**Comment by Lou1415926** [comment]
oohhh! Thanks for the clarification!

> If it mattered that I am not using CodeCommit to host my repository, I am hosting the submodule on GitHub. Does that mean I need to set a different permission?

If your submodule is hosted on GitHub, then the `url` provided to `git config` should probably be the GitHub URL. I am guessing #2799 used that "git-codecommit" URL probably because they hosted their submodules in CodeCommit instead of GitHub. 

Given the information that you provided, I feel like [this page](https://medium.com/@cristiano.ventura/working-with-git-submodules-in-codepipeline-83e843e5d0a) and [this page](https://stackoverflow.com/questions/42712542/how-to-auto-deploying-git-repositories-with-submodules-on-aws) are better suited for you! Basically, you will need to configure your GitHub credential so that you can pull the codes from your submodule repository.

Please let me know if this works for you! I'm happy to follow up and help!

P.S.: I gave the answer above- that is, adding `codecommit::GitPull` - because I thought your submodule code was hosted in a CodeCommit repository. If that's not the case, you don't need that permission!

**Comment by dmathewwws** [comment]
👌🏾Worked perfectly. Thanks!

For anyone else. I followed the instructions from `imekinox` for https://stackoverflow.com/a/64163023.

**Comment by surrealchemist** [comment]
I just wanted to mention I had been using the instructions in that stack overflow in the past, but while setting up a new pipeline with submodules today I found the entry in the documentation source.properties.[output_artifact_format](https://aws.github.io/copilot-cli/docs/manifest/pipeline/#source-properties-output-artifact-format) and you just have to add that to your pipeline manifest setting it to CODEBUILD_CLONE_REF. 

So basically to deploy a new pipeline (for github anyway):
- Run the pipeline init command
- Modify the pipeline buildspec for the submodule update line
- Modify the pipeline manifest to add the clone ref option
- Commit the pipeline files and deploy the pipeline
